File: CI/check-mpackage-sync.py
#!/usr/bin/env python3
"""Check packaged .mpackage archives against their checked-in sources.

Mudlet installs the .mpackage archive, not the loose config.lua/.xml files
sitting next to it, so editing a source file without rebuilding the archive
silently changes nothing at all.

Some of these packages are also published to the package repository
(Mudlet/mudlet-package-repository), which offers updates by comparing the
version in config.lua. A content change that keeps the old version number
never reaches players who installed the package with mpkg.

Run with no arguments to check archive contents. Pass --base-ref to also
require a version bump for any package whose contents changed.
"""

import argparse
import io
import subprocess
import sys
import zipfile
from pathlib import Path

# packages built from the loose sources next to them - the two must agree
SOURCED_PACKAGES = [
    "src/mudlet-lua/lua/base-ui/mudlet-base-ui.mpackage",
    "src/mudlet-lua/lua/generic-mapper/generic_mapper.mpackage",
    "src/mudlet-lua/lua/gui-drop/gui-drop.mpackage",
]

# packages the repository syncs weekly, where mpkg needs a version bump to
# offer the update - see update-core-packages.yml in the package repository
PUBLISHED_PACKAGES = [
    "src/deleteOldProfiles.mpackage",
    "src/echo.mpackage",
    "src/enable-accessibility.mpackage",
    "src/mudlet-lua/lua/base-ui/mudlet-base-ui.mpackage",
    "src/mudlet-lua/lua/generic-mapper/generic_mapper.mpackage",
    "src/run-lua-code.mpackage",
]

errors = []
warnings = []


def contents(archive):
    """Map member name to bytes, ignoring directory entries."""
    with zipfile.ZipFile(archive) as package:
        return {member.filename: package.read(member) for member in package.infolist() if not member.is_dir()}


def version_of(members):
    config = members.get("config.lua", b"").decode("utf-8", "replace")
    for line in config.splitlines():
        name, _, value = line.partition("=")
        if name.strip() == "version":
            return value.strip().strip("[]\"' ")
    return None


def as_number(version):
    """Sortable form of a version, tolerating parts like "2" or "1.0.0rc1"."""
    return [(int("".join(filter(str.isdigit, part)) or 0), part) for part in version.split(".")]


def at_base_ref(path, base_ref):
    """Archive contents at base_ref, or None if the package is new there."""
    result = subprocess.run(["git", "show", f"{base_ref}:{path}"], capture_output=True)
    if result.returncode != 0:
        return None
    return contents(io.BytesIO(result.stdout))


def check_sources_match(path, members, enforced):
    """Every member with a file of the same name beside the archive must match it."""
    for name, packaged in members.items():
        source = Path(path).parent / name
        if not source.is_file():
            continue
        if source.read_bytes() == packaged:
            continue
        complaint = f"{path} does not match {source} - rebuild the archive after editing the source"
        (errors if enforced else warnings).append(complaint)


def check_version_bumped(path, members, base_ref):
    was = at_base_ref(path, base_ref)
    if was is None or was == members:
        return

    old, new = version_of(was), version_of(members)
    if not new:
        errors.append(f"{path} has no version in its config.lua")
    elif old and as_number(new) <= as_number(old):
        errors.append(f"{path} changed but is still version {new} - bump it so mpkg offers the update")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--base-ref", help="branch to compare against, e.g. origin/development")
    arguments = parser.parse_args()

    for path in sorted(set(SOURCED_PACKAGES) | set(PUBLISHED_PACKAGES)):
        if not Path(path).is_file():
            errors.append(f"{path} is listed in {__file__} but does not exist")
            continue

        members = contents(path)
        check_sources_match(path, members, enforced=path in SOURCED_PACKAGES)
        if arguments.base_ref and path in PUBLISHED_PACKAGES:
            check_version_bumped(path, members, arguments.base_ref)

    for warning in warnings:
        print(f"warning: {warning}")
    for error in errors:
        print(f"error: {error}")

    if errors:
        print(f"\n{len(errors)} problem(s) found. Rebuild an archive with:")
        print("  cd <package directory> && zip -r <name>.mpackage config.lua <name>.xml .mudlet")
        return 1

    print("mpackage archives match their sources.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: CI/check-mpackage-sync.lua
#!/usr/bin/env lua
--[[
Check packaged .mpackage archives against their checked-in sources.

Mudlet installs the .mpackage archive, not the loose config.lua/.xml files
sitting next to it, so editing a source file without rebuilding the archive
silently changes nothing at all.

Some of these packages are also published to the package repository
(Mudlet/mudlet-package-repository), which offers updates by comparing the
version in config.lua. A content change that keeps the old version number
never reaches players who installed the package with mpkg.

Run with no arguments to check archive contents. Pass --base-ref to also
require a version bump for any package whose contents changed:

  lua CI/check-mpackage-sync.lua --base-ref origin/development
]]

-- packages the package repository syncs weekly, where mpkg needs a version bump
-- to offer the update - see update-core-packages.yml over there
local publishedPackages = {
  "src/packages/deleteOldProfiles/deleteOldProfiles.mpackage",
  "src/packages/echo/echo.mpackage",
  "src/packages/enable-accessibility/enable-accessibility.mpackage",
  "src/packages/generic_mapper/generic_mapper.mpackage",
  "src/packages/mudlet-base-ui/mudlet-base-ui.mpackage",
  "src/packages/run-lua-code/run-lua-code.mpackage",
}

local errors = {}

local function contains(list, wanted)
  for _, item in ipairs(list) do
    if item == wanted then return true end
  end
  return false
end

local function quote(argument)
  return "'" .. argument:gsub("'", "'\\''") .. "'"
end

local function capture(command)
  local pipe = assert(io.popen(command, "r"))
  local output = pipe:read("*a")
  pipe:close()
  return output
end

local function readFile(path)
  local file = io.open(path, "rb")
  if not file then return nil end
  local contents = file:read("*a")
  file:close()
  return contents
end

-- Every entry in the archive, mapped to its bytes. Directory entries, which
-- zipinfo lists with a trailing slash, are not files and are skipped.
local function contentsOf(archive)
  local members = {}
  for name in capture("unzip -Z1 " .. quote(archive)):gmatch("[^\n]+") do
    if not name:match("/$") then
      members[name] = capture(string.format("unzip -p %s %s", quote(archive), quote(name)))
    end
  end
  return members
end

local function sameContents(one, other)
  for name, bytes in pairs(one) do
    if other[name] ~= bytes then return false end
  end
  for name in pairs(other) do
    if one[name] == nil then return false end
  end
  return true
end

local function versionOf(members)
  for line in (members["config.lua"] or ""):gmatch("[^\n]+") do
    local version = line:match("^version%s*=%s*(.-)%s*$")
    if version then
      return (version:gsub("^[%[\"']+", ""):gsub("[%]\"']+$", ""))
    end
  end
  return nil
end

-- Sortable form of a version, tolerating parts like "2" or "1.0.0rc1"
local function versionParts(version)
  local parts = {}
  for part in version:gmatch("[^.]+") do
    parts[#parts + 1] = {tonumber(part:match("%d+")) or 0, part}
  end
  return parts
end

local function isNewer(candidate, existing)
  local new, old = versionParts(candidate), versionParts(existing)
  for index = 1, math.max(#new, #old) do
    local newPart = new[index] or {0, ""}
    local oldPart = old[index] or {0, ""}
    if newPart[1] ~= oldPart[1] then return newPart[1] > oldPart[1] end
    if newPart[2] ~= oldPart[2] then return newPart[2] > oldPart[2] end
  end
  return false
end

-- Archive contents at baseRef, or nil if the package is new there
local function contentsAtBaseRef(path, baseRef)
  local temporary = os.tmpname()
  local archive = capture(string.format("git show %s 2>/dev/null", quote(baseRef .. ":" .. path)))
  if archive == "" then
    os.remove(temporary)
    return nil
  end

  local file = assert(io.open(temporary, "wb"))
  file:write(archive)
  file:close()
  local members = contentsOf(temporary)
  os.remove(temporary)
  return members
end

-- Every member with a file of the same name beside the archive must match it
local function checkSourcesMatch(path, members)
  local directory = path:match("^(.*)/[^/]+$")
  for name, packaged in pairs(members) do
    local source = directory .. "/" .. name
    local onDisk = readFile(source)
    if onDisk and onDisk ~= packaged then
      table.insert(errors, string.format("%s does not match %s - rebuild the archive after editing the source", path, source))
    end
  end
end

local function checkVersionBumped(path, members, baseRef)
  local was = contentsAtBaseRef(path, baseRef)
  if not was or sameContents(was, members) then return end

  local old, new = versionOf(was), versionOf(members)
  if not new then
    table.insert(errors, string.format("%s has no version in its config.lua", path))
  elseif old and not isNewer(new, old) then
    table.insert(errors, string.format("%s changed but is still version %s - bump it so mpkg offers the update", path, new))
  end
end

local baseRef
for index = 1, #arg do
  if arg[index] == "--base-ref" then
    baseRef = arg[index + 1]
  elseif arg[index]:match("^%-%-base%-ref=") then
    baseRef = arg[index]:match("=(.*)$")
  end
end

-- every default package lives in its own directory under src/packages, named
-- after the package, holding the archive and the sources it was built from
local packages = {}
for name in capture("ls -1 src/packages"):gmatch("[^\n]+") do
  local archive = string.format("src/packages/%s/%s.mpackage", name, name)
  if readFile(archive) then table.insert(packages, archive) end
end
table.sort(packages)

for _, path in ipairs(publishedPackages) do
  if not contains(packages, path) then
    table.insert(errors, string.format("%s is listed as published but is not in src/packages", path))
  end
end

for _, path in ipairs(packages) do
  local members = contentsOf(path)
  checkSourcesMatch(path, members)
  if baseRef and contains(publishedPackages, path) then
    checkVersionBumped(path, members, baseRef)
  end
end

for _, message in ipairs(errors) do
  print("error: " .. message)
end

if #errors > 0 then
  print(string.format("\n%d problem(s) found. Rebuild an archive from its sources with:", #errors))
  print("  cd src/packages/<name> && zip <name>.mpackage config.lua <name>.xml")
  os.exit(1)
end

print(string.format("%d mpackage archives match their sources.", #packages))
